File: src/main/java/ttd/site/generator/BlogItemWriter.java
package ttd.site.generator;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.batch.item.ItemWriter;
import org.springframework.util.ReflectionUtils;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author <a href="mailto:josh@joshlong.com">Josh Long</a>
 */
@Log4j2
@RequiredArgsConstructor
class BlogItemWriter implements ItemWriter<Map<String, Collection<Bookmark>>> {

	private final TemplateService templateService;

	private final SiteGeneratorConfigurationProperties properties;

	private void writeBlog(String publishKey, Collection<Bookmark> bookmarks) {
		var date = DateUtils.parseYearMonthDay(publishKey);
		var pk = publishKey.trim();
		var file = new File(this.properties.getContentDirectory(), pk + ".html");
		var links = bookmarks.stream()
				.map(bm -> new Link(Long.toString(bm.getBookmarkId()), bm.getHref(), bm.getDescription(), bm.getTime()))
				.collect(Collectors.toSet());
		try (var bw = new BufferedWriter(new FileWriter(file))) {
			bw.write(this.templateService.daily(date, links));
		}
		catch (IOException e) {
			ReflectionUtils.rethrowRuntimeException(e);
		}
	}

	@Override
	public void write(List<? extends Map<String, Collection<Bookmark>>> list) {
		list.forEach((map) -> map.forEach(BlogItemWriter.this::writeBlog));
	}

}

File: src/main/kotlin/ttd/site/generator/BlogItemWriter.kt
package ttd.site.generator

import org.springframework.batch.item.ItemWriter
import org.springframework.util.ReflectionUtils
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.util.stream.Collectors

class BlogItemWriter(
		private val templateService: TemplateService,
		private val properties: SiteGeneratorConfigurationProperties) : ItemWriter<Map<String, Collection<Bookmark>>> {

	private fun writeBlog(publishKey: String, bookmarks: Collection<Bookmark>) {
		val date = DateUtils.parseYearMonthDay(publishKey)
		val pk = publishKey.trim { it <= ' ' }
		val file = File(properties.contentDirectory.file, "$pk.html")
		val links = bookmarks.stream()
				.map { bm: Bookmark -> Link(bm.bookmarkId.toString(), bm.href, bm.description, bm.time) }
				.collect(Collectors.toSet())
		try {
			BufferedWriter(FileWriter(file)).use { bw -> bw.write(templateService.daily(date, links)) }
		} catch (e: IOException) {
			ReflectionUtils.rethrowRuntimeException(e)
		}
	}

	override fun write(items: List<Map<String, Collection<Bookmark>>>) {
		items.forEach { map ->
			map.forEach { (publishKey, bookmarks) ->
				writeBlog(publishKey, bookmarks)
			}
		}
	}
}